src/components/external-apis/routes.ts

```diff
@@ -0,0 +1,130 @@
+import { FastifyRequest, FastifyReply } from 'fastify'
+import path from 'path'
+import config from '../../config'
+import { get as extractProperty } from 'lodash'
+import { getPermissionNamesFromJWT } from '../data_display/helpers'
+import axios, { AxiosError, AxiosRequestConfig } from 'axios'
+import { constructAuthHeader, constructQueryObject, validateResult } from './helpers'
+import { ExternalApiConfigs, QueryParameters, PostRoute } from './types'
+import { getApplicationData } from '../actions'
+import { getUserInfo } from '../permissions/loginHelpers'
+import { ActionApplicationData } from '../../types'
+import functions from '../actions/evaluatorFunctions'
+import { errorMessage } from '../utilityFunctions'
+
+interface ExternalApiRequest {
+  name: string
+  route: string
+}
+
+const apiConfigs: ExternalApiConfigs = config?.externalApiConfigs ?? {}
+
+export const routeAccessExternalApi = async (
+  request: FastifyRequest & { auth?: { userId?: number; orgId?: number } },
+  reply: FastifyReply
+) => {
+  const { name, route } = request.params as ExternalApiRequest
+
+  const { baseUrl, routes, authentication } = apiConfigs?.[name]
+  if (!baseUrl) {
+    throw new Error('No baseUrl defined for API: ' + name)
+  }
+
+  const routeConfig = routes?.[route]
+  if (!routeConfig) {
+    throw new Error('No config definition for Route: ' + route)
+  }
+
+  const {
+    method,
+    url,
+    permissions,
+    queryParams,
+    allowedClientQueries,
+    returnProperty,
+    additionalAxiosProperties,
+    validationExpression,
+  } = routeConfig
+
+  if (permissions) {
+    const { permissionNames } = await getPermissionNamesFromJWT(request)
+    const validPermissions = permissions.filter((permission) =>
```

Comment:
should all permissions be validated ?

const notMatchedPermission = permissions.find(permission) => !permissioneName.includes(permission))

Reply:
I'm not quite clear what you're suggesting here. Do you mean that the user should have *all* of the specified permissions? I was thinking *any* of the specified permissions would be fine (so *OR* not *AND*), for 3 reasons:
- this is consistent with how it works for Data Views (if you have any of the listed permissions you can view)
- it seems to make more sense that we'd be defining in terms of groups of people who can view something. For example, if we had the `applyCompany` permission required, and then we wanted to include `admin` so Admins could do it too, then we wouldn't want suddenly exclude company applicants.

Or am I missing what you meant?

Reply:
Ahh ok, sorry i missed: "if defined, the client request must have one of these permissions (in its JWT token) in order to proceed"